set up CI

noot opened the pull request:
attempting to set up a CI for building + testing (and maybe fmt-ing :P)

noot committed:
begin to setup ci

noot committed:
attempt to fix build

noot committed:
fix paths in build script

noot committed:
fix

noot committed:
satisfy clippy

noot committed:
update fmt check to use nightly

noot committed:
merge w develop

kayabaNerve commented on .github/workflows/tests.yml:
```diff
@@ -0,0 +1,38 @@
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v1
+      - run: rustup component add rustfmt
```
nightly fmt

noot replied:
yep added

kayabaNerve commented on scripts/test.sh:
```diff
@@ -0,0 +1,3 @@
+#!/bin/bash
+./coins/monero/c/monero/build/release/bin/monerod --detach --regtest --offline --fixed-difficulty=1 --rpc-bind-ip 127.0.0.1 --rpc-bind-port 18081
+cargo test
```
`--all-features` ?

kayabaNerve replied:
Looks like this isn't feasible because try-runtime doesn't work for the current Substrate runtime. One error seems to be a macro not realizing what it's doing, the others... I'm not sure about. We could remove try-runtime for now?

kayabaNerve commented on scripts/build.sh:
```diff
@@ -0,0 +1,8 @@
+#!/bin/bash
+BASEDIR=$(pwd)
+git submodule update --init --recursive
+sudo apt-get install libgtest-dev && cd /usr/src/gtest && sudo cmake . && sudo make
```
I don't think this is necessary for Monero, just requested. We can slim this down a good bit (notably the following apt) by removing optionals.

noot replied:
that would be nice, i just copy-pasted this from the Monero readme lol

silverpill replied:
This is the list of dependencies I used for building the project on Debian:

```
apt install cmake libzmq3-dev libboost-dev libboost-system-dev libboost-filesystem-dev libboost-thread-dev libboost-date-time-dev libboost-chrono-dev libboost-regex-dev libboost-serialization-dev libboost-program-options-dev libboost-locale-dev
```

May be sufficient for Ubuntu too since it's Debian-based.

Also, last time I checked Serai required Rust 1.60 stable.

kayabaNerve replied:
Thanks for chiming in with that :)

I'm actually not sure what version of Rust we require + why. It may be good to go through each crate, working that, and adding `msrv`s accordingly, working on finding earlier versions which are trivial to move to in order to bring wide support.

silverpill replied:
I checked out the latest commit and there were 458 new crates to download :exploding_head: 
- `librocksdb-sys` requires `apt install clang`
- `ethereum-serai` requires solc binary (I think it can be installed from PPA on Ubuntu)
- `serai-runtime` requires rust-nightly (in build script)
- `monero-serai` fails with ```no method named `len` found for type `usize` in the current scope``` <del>(I'm not sure if it has anything to do with rust version)</del> https://github.com/serai-dex/serai/pull/51
- Apparently the rest of dependency tree still builds with 1.60, though I didn't get to the end of it

Rust 1.60 requirement probably comes from namespaced features in [some manifests](https://github.com/serai-dex/serai/blob/develop/crypto/transcript/Cargo.toml#L19) - that was the error I encountered when I tried to build project with 1.56 initially.

kayabaNerve replied:
The new crates would be the mess that is Substrate.

solc should be from solc-select, which is a pip3.

monero-serai is absolutely my fault and dates back to when I cleaned up CLSAG a bit due to cargo fmt's opinion on it.

Anything wasm under Substrate will require nightly :/

Thanks for clarifying there. I forgot Rust only recently got sane feature naming...

noot committed:
use nightly for build

kayabaNerve commented on .github/workflows/tests.yml:
```diff
@@ -0,0 +1,38 @@
+
+on: [pull_request]
+
+jobs:
```
Also needs a cache.

https://github.com/MerosCrypto/asmr/blob/master/.github/workflows/ci.yaml#L54-L84 for an example

noot committed:
fmt

noot committed:
fix fmt install

noot committed:
update test script

noot committed:
try to fix fmt

noot committed:
Merge branch 'develop' of github.com:serai-dex/serai into noot/setup-ci

noot committed:
merge w develop

noot committed:
Merge branch 'develop' of github.com:serai-dex/serai into noot/setup-ci

noot committed:
maybe fix build script

noot committed:
install wasm toolchain

kayabaNerve commented on .github/workflows/tests.yml:
```diff
@@ -10,6 +10,8 @@ jobs:
     - name: Install nightly
       run: rustup toolchain install nightly
+    - name: Install WASM toolchain
+      run: rustup target add wasm32-unknown-unknown
```
stable rust, nightly wasm

noot committed:
install solc-select, use stable rust to build

kayabaNerve committed:
Merge branch 'develop' into noot/setup-ci

kayabaNerve committed:
Correct clippy warnings

Currently intended to be done with:
cargo clippy --features "recommended merlin batch serialize experimental 
ed25519 ristretto p256 secp256k1 multisig" -- -A clippy::type_complexity 
-A dead_code

kayabaNerve committed:
Remove try-runtime

I tried to get this to work for an hour. I have no idea why it doesn't, 
yet it doesn't.

kayabaNerve committed:
Rewrite workflow

Splits tasks into a more modular structure. Also uses 
actions-rs/toolchain.

kayabaNerve committed:
Add a cache

kayabaNerve committed:
Immediately try building ETH/Monero while this is fixed

Adds solc-select use.

kayabaNerve committed:
Revert selective advance building of ETH/XMR

ETH builds now, so it hopefully should work now.

Also moves from on push to on push to develop.

kayabaNerve committed:
Install Monero runtime dependencies

Specify missing Rust toolchain setting.

kayabaNerve committed:
Correct multi-line commands

kayabaNerve committed:
Fix multi-line commands again

Cache Ethereum artifacts.

kayabaNerve committed:
Add Foundry

kayabaNerve committed:
Move Clippy under build

kayabaNerve committed:
Minimal rustup

Adds wasm Clippy. Puts Clippy before build.

kayabaNerve committed:
Use nightly clippy

kayabaNerve committed:
Remove old clippy call from under build

kayabaNerve committed:
Have the Monero build script support ARCH specification

Requirement for CI.

kayabaNerve committed:
Add WASM toolchain to tests

kayabaNerve committed:
Remove Ethereum cache which did not work as needed

kayabaNerve committed:
Remove extraneous quotes which broke builds on Arch
